File: OurAdtA01MessageBuilder.java
package com.saravanansubramanian.hapihl7tutorial;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;

import ca.uhn.hl7v2.HL7Exception;
import ca.uhn.hl7v2.model.DataTypeException;
import ca.uhn.hl7v2.model.v24.datatype.PL;
import ca.uhn.hl7v2.model.v24.datatype.XAD;
import ca.uhn.hl7v2.model.v24.datatype.XCN;
import ca.uhn.hl7v2.model.v24.datatype.XPN;
import ca.uhn.hl7v2.model.v24.message.ADT_A01;
import ca.uhn.hl7v2.model.v24.segment.EVN;
import ca.uhn.hl7v2.model.v24.segment.MSH;
import ca.uhn.hl7v2.model.v24.segment.PID;
import ca.uhn.hl7v2.model.v24.segment.PV1;

public class OurAdtA01MessageBuilder {

	private ADT_A01 _adtMessage;

	/*
	 * You can pass in a domain object as a parameter when integrating with data
	 * from your application here and I will leave that to you to explore on your
	 * own. I will use fictional data here for illustration
	 */

	public ADT_A01 Build() throws HL7Exception, IOException {
		String currentDateTimeString = getCurrentTimeStamp();
		_adtMessage = new ADT_A01();
		_adtMessage.initQuickstart("ADT", "A01", "P");
		createMshSegment(currentDateTimeString);
		createEvnSegment(currentDateTimeString);
		createPidSegment();
		createPv1Segment();
		return _adtMessage;
	}

	private void createMshSegment(String currentDateTimeString) throws DataTypeException {
		MSH mshSegment = _adtMessage.getMSH();
		mshSegment.getFieldSeparator().setValue("|");
		mshSegment.getEncodingCharacters().setValue("^~\\&");
		mshSegment.getSendingApplication().getNamespaceID().setValue("Our System");
		mshSegment.getSendingFacility().getNamespaceID().setValue("Our Facility");
		mshSegment.getReceivingApplication().getNamespaceID().setValue("Their Remote System");
		mshSegment.getReceivingFacility().getNamespaceID().setValue("Their Remote Facility");
		mshSegment.getDateTimeOfMessage().getTimeOfAnEvent().setValue(currentDateTimeString);
		mshSegment.getMessageControlID().setValue(getSequenceNumber());
		mshSegment.getVersionID().getVersionID().setValue("2.4");
	}

	private void createEvnSegment(String currentDateTimeString) throws DataTypeException {
		EVN evn = _adtMessage.getEVN();
		evn.getEventTypeCode().setValue("A01");
		evn.getRecordedDateTime().getTimeOfAnEvent().setValue(currentDateTimeString);
	}

	private void createPidSegment() throws DataTypeException {
		PID pid = _adtMessage.getPID();
		XPN patientName = pid.getPatientName(0);
		patientName.getFamilyName().getSurname().setValue("Mouse");
		patientName.getGivenName().setValue("Mickey");
		pid.getPatientIdentifierList(0).getID().setValue("378785433211");
		XAD patientAddress = pid.getPatientAddress(0);
		patientAddress.getStreetAddress().getStreetOrMailingAddress().setValue("123 Main Street");
		patientAddress.getCity().setValue("Lake Buena Vista");
		patientAddress.getStateOrProvince().setValue("FL");
		patientAddress.getCountry().setValue("USA");
	}

	private void createPv1Segment() throws DataTypeException {
		PV1 pv1 = _adtMessage.getPV1();
		pv1.getPatientClass().setValue("O"); // to represent an 'Outpatient'
		PL assignedPatientLocation = pv1.getAssignedPatientLocation();
		assignedPatientLocation.getFacility().getNamespaceID().setValue("Some Treatment Facility Name");
		assignedPatientLocation.getPointOfCare().setValue("Some Point of Care");
		pv1.getAdmissionType().setValue("ALERT");
		XCN referringDoctor = pv1.getReferringDoctor(0);
		referringDoctor.getIDNumber().setValue("99999999");
		referringDoctor.getFamilyName().getSurname().setValue("Smith");
		referringDoctor.getGivenName().setValue("Jack");
		referringDoctor.getIdentifierTypeCode().setValue("456789");
		pv1.getAdmitDateTime().getTimeOfAnEvent().setValue(getCurrentTimeStamp());
	}

	private String getCurrentTimeStamp() {
		return new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
	}

	private String getSequenceNumber() {
		String facilityNumberPrefix = "1234"; // some arbitrary prefix for the facility
		return facilityNumberPrefix.concat(getCurrentTimeStamp());
	}
}

File: HL7 Programming using DotNet - Tutorial Series/Creating HL7 Messages/NHapiCreateMessageSimpleExample/OurAdtA01MessageBuilder.cs
using System;
using System.Globalization;
using NHapi.Model.V23.Message;

namespace NHapiCreateMessageSimpleExample
{
    internal class OurAdtA01MessageBuilder
    {
        private ADT_A01 _adtMessage;

        /*You can pass in a domain or data transfer object as a parameter
        when integrating with data from your application here
        I will leave that to you to explore on your own
        Using fictional data here for illustration*/

        public ADT_A01 Build()
        {
            var currentDateTimeString = GetCurrentTimeStamp();
            _adtMessage = new ADT_A01();
            
            CreateMshSegment(currentDateTimeString);
            CreateEvnSegment(currentDateTimeString);
            CreatePidSegment();
            CreatePv1Segment();
            return _adtMessage;
        }

        private void CreateMshSegment(string currentDateTimeString) 
        {
            var mshSegment = _adtMessage.MSH;
            mshSegment.FieldSeparator.Value = "|";
            mshSegment.EncodingCharacters.Value = "^~\\&";
            mshSegment.SendingApplication.NamespaceID.Value = "Our System";
            mshSegment.SendingFacility.NamespaceID.Value = "Our Facility";
            mshSegment.ReceivingApplication.NamespaceID.Value = "Their Remote System";
            mshSegment.ReceivingFacility.NamespaceID.Value = "Their Remote Facility";
            mshSegment.DateTimeOfMessage.TimeOfAnEvent.Value = currentDateTimeString;
            mshSegment.MessageControlID.Value = GetSequenceNumber();
            mshSegment.MessageType.MessageType.Value = "ADT";
            mshSegment.MessageType.TriggerEvent.Value = "A01";
            mshSegment.VersionID.Value = "2.3";
            mshSegment.ProcessingID.ProcessingID.Value = "P";
        }

        private void CreateEvnSegment(string currentDateTimeString) 
        {
            var evn = _adtMessage.EVN;
            evn.EventTypeCode.Value = "A01";
            evn.RecordedDateTime.TimeOfAnEvent.Value = currentDateTimeString;
        }

        private void CreatePidSegment() 
        {
            var pid = _adtMessage.PID;
            var patientName = pid.GetPatientName(0);
            patientName.FamilyName.Value = "Mouse";
            patientName.GivenName.Value = "Mickey";
            pid.SetIDPatientID.Value = "378785433211";
            var patientAddress = pid.GetPatientAddress(0);
            patientAddress.StreetAddress.Value = "123 Main Street";
            patientAddress.City.Value = "Lake Buena Vista";
            patientAddress.StateOrProvince.Value = "FL";
            patientAddress.Country.Value = "USA";
        }

        private void CreatePv1Segment() 
        {
            var pv1 = _adtMessage.PV1;
            pv1.PatientClass.Value = "O"; // to represent an 'Outpatient'
            var assignedPatientLocation = pv1.AssignedPatientLocation;
            assignedPatientLocation.Facility.NamespaceID.Value = "Some Treatment Facility";
            assignedPatientLocation.PointOfCare.Value = "Some Point of Care";
            pv1.AdmissionType.Value = "ALERT";
            var referringDoctor = pv1.GetReferringDoctor(0);
            referringDoctor.IDNumber.Value = "99999999";
            referringDoctor.FamilyName.Value = "Smith";
            referringDoctor.GivenName.Value = "Jack";
            referringDoctor.IdentifierTypeCode.Value = "456789";
            pv1.AdmitDateTime.TimeOfAnEvent.Value = GetCurrentTimeStamp();
        }

        private static string GetCurrentTimeStamp()
        {
            return DateTime.Now.ToString("yyyyMMddHHmmss",CultureInfo.InvariantCulture);
        }

        private static string GetSequenceNumber()
        {
            const string facilityNumberPrefix = "1234"; // some arbitrary prefix for the facility
            return facilityNumberPrefix + GetCurrentTimeStamp();
        }
    }
}